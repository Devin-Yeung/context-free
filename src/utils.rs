use bnf::{Grammar, Term};
use std::collections::HashSet;

pub fn symbols(grammar: &Grammar) -> HashSet<&Term> {
    grammar
        .productions_iter()
        .flat_map(|production| production.rhs_iter())
        .flat_map(|expr| expr.terms_iter())
        .collect::<HashSet<_>>()
}
