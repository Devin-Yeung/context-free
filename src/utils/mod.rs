use bnf::{Grammar, Term};
use once_cell::sync::OnceCell;
use std::collections::HashSet;
use std::hash::Hash;
use std::path::Iter;

pub mod first;
pub mod follow;

pub fn symbols(grammar: &Grammar) -> HashSet<&Term> {
    grammar
        .productions_iter()
        .flat_map(|production| {
            production
                .rhs_iter()
                .flat_map(|expr| expr.terms_iter())
                .chain(std::iter::once(&production.lhs))
        })
        .collect::<HashSet<_>>()
}

pub fn terminals(grammar: &Grammar) -> impl Iterator<Item = &Term> {
    symbols(&grammar)
        .into_iter()
        .filter(|term| term != &epsilon())
        .filter(|term| matches!(*term, Term::Terminal(_)))
}

pub fn nonterminals(grammar: &Grammar) -> impl Iterator<Item = &Term> {
    symbols(&grammar)
        .into_iter()
        .filter(|term| term != &epsilon())
        .filter(|term| matches!(*term, Term::Nonterminal(_)))
}

pub fn epsilon() -> &'static Term {
    static EPSILON: OnceCell<Term> = OnceCell::new();
    EPSILON.get_or_init(|| Term::Terminal(String::from("ε")))
}

pub fn dollar() -> &'static Term {
    static DOLLAR: OnceCell<Term> = OnceCell::new();
    DOLLAR.get_or_init(|| Term::Terminal(String::from("$")))
}

#[cfg(test)]
mod tests {
    use super::symbols;

    #[test]
    fn count_symbols() {
        let grammar = r#"
        <P> ::= <Q> 'id' <R>
        <Q> ::= '∃' | '∀'
        <R> ::= <E> '=' <E>
        <E> ::= <E> '+' <T> | <T>
        <T> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();

        assert_eq!(symbols(&grammar).len(), 12);
    }
}
