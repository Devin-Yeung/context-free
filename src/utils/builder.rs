use crate::utils::symbols;
use bnf::{Grammar, Term};
use once_cell::sync::OnceCell;
use std::collections::{HashMap, HashSet};

pub fn epsilon() -> &'static Term {
    static EPSILON: OnceCell<Term> = OnceCell::new();
    EPSILON.get_or_init(|| Term::Terminal(String::from("ε")))
}

fn dollar() -> &'static Term {
    static DOLLAR: OnceCell<Term> = OnceCell::new();
    DOLLAR.get_or_init(|| Term::Terminal(String::from("$")))
}

pub struct FirstFollowBuilder<'grammar> {
    pub(crate) inner: HashMap<&'grammar Term, HashSet<&'grammar Term>>,
}

impl<'grammar> FirstFollowBuilder<'grammar> {
    pub(crate) fn new(grammar: &'grammar Grammar) -> FirstFollowBuilder<'grammar> {
        let mut inner = HashMap::new();

        // initialize the table
        symbols(&grammar)
            .into_iter()
            .filter(|term| term != &epsilon()) // epsilon is a special non-terminal
            .for_each(|term| {
                inner.insert(term, HashSet::new());
            });

        FirstFollowBuilder { inner }
    }

    // Insert term to First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    pub(crate) fn insert_term(&mut self, x: &'grammar Term, term: &'grammar Term) -> bool {
        // First(x)
        let first_x = self.inner.get_mut(x).unwrap();

        // Insert term to First(x)
        let before = first_x.len();
        first_x.insert(term);
        let after = first_x.len();

        // check if set changes
        before != after
    }

    // Insert epsilon to First(x)
    pub(crate) fn insert_epsilon(&mut self, x: &'grammar Term) -> bool {
        self.insert_term(x, epsilon())
    }

    // Insert dollar to Follow(x)
    pub(crate) fn insert_dollar(&mut self, x: &'grammar Term) -> bool {
        self.insert_term(x, dollar())
    }

    /// First(x)
    pub(crate) fn first(&self, x: &Term) -> HashSet<&'grammar Term> {
        self.inner
            .get(x)
            .map_or_else(|| HashSet::new(), |set| set.clone())
    }

    /// Follow(x)
    pub(crate) fn follow(&self, x: &Term) -> HashSet<&'grammar Term> {
        self.inner
            .get(x)
            .map_or_else(|| HashSet::new(), |set| set.clone())
    }

    pub(crate) fn insert_set(&mut self, x: &'grammar Term, set: HashSet<&'grammar Term>) -> bool {
        // First(x)
        let first_x = self.inner.get_mut(x).unwrap();

        // Insert set into First(x)
        let before = first_x.len();
        first_x.extend(set);
        let after = first_x.len();

        // check if set changes
        return before != after;
    }

    /// Insert First(y) \ { ε } into First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    pub(crate) fn insert_first_no_epsilon(&mut self, x: &'grammar Term, y: &'grammar Term) -> bool {
        // First(y)
        let mut first_y = self.first(y);
        // First(y) \ { ε }
        first_y.remove(epsilon());
        // Insert First(y) \ { ε } into First(x)
        self.insert_set(x, first_y)
    }

    /// Insert Follow(y) into Follow(x)
    pub(crate) fn insert_follow(&mut self, x: &'grammar Term, y: &'grammar Term) -> bool {
        // Follow(y)
        let follow_y = self.follow(y);
        println!("Insert {:?} to Follow({})", follow_y, x);
        // Insert Follow(y) into Follow(x)
        self.insert_set(x, follow_y)
    }

    pub(crate) fn build(self) -> HashMap<&'grammar Term, HashSet<&'grammar Term>> {
        self.inner
    }
}
