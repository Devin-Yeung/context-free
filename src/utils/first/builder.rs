use crate::utils::first::First;
use crate::utils::{epsilon, symbols};
use bnf::{Grammar, Production, Term};
use std::cell::RefCell;
use std::collections::{HashMap, HashSet};

pub struct FirstBuilder<'grammar> {
    pub(crate) grammar: &'grammar Grammar,
    pub(crate) first: RefCell<HashMap<&'grammar Term, HashSet<&'grammar Term>>>,
    pub(crate) lookup: HashMap<&'grammar Term, &'grammar Production>,
}

impl<'grammar> FirstBuilder<'grammar> {
    pub(crate) fn new(grammar: &'grammar Grammar) -> FirstBuilder<'grammar> {
        let mut first = HashMap::new();

        let lookup = grammar
            .productions_iter()
            .map(|production| (&production.lhs, production))
            .collect::<HashMap<_, _>>();

        // initialize the table
        symbols(grammar)
            .filter(|term| term != &epsilon()) // epsilon is a special non-terminal
            .for_each(|term| {
                first.insert(term, HashSet::new());
            });

        let first = RefCell::new(first);

        FirstBuilder {
            grammar,
            first,
            lookup,
        }
    }

    fn build_first(&mut self) {
        symbols(self.grammar)
            .filter(|term| term != &epsilon())
            .for_each(|t| {
                match t {
                    Term::Terminal(s) => {
                        // Rule1: If X is a terminal, then First(X) = { X }
                        self.insert_term(t, t);
                        println!("Rule1: Push {} to First({})", s, t);
                    }
                    Term::Nonterminal(_) => { /* skip */ }
                };

                if self.produce_epsilon(t) {
                    // Rule2: If X is an ε-production, then add ε to First(X)
                    self.insert_epsilon(t);
                    println!("Rule2: Push ε to First({})", t);
                }
            });

        loop {
            let mut changed = false;

            symbols(self.grammar)
                .filter(|term| matches!(*term, Term::Nonterminal(_)))
                .for_each(|lhs| {
                    println!("===> Checking Symbol: {}", lhs);
                    let production = self.lookup.get(lhs).unwrap();
                    // Rule3: If X is a non-terminal and X → Y1 Y2 ... Yk,
                    // then add First(Y1) ∖ {ε} to First(X)
                    for expr in production.rhs_iter() {
                        for term in expr
                            .terms_iter()
                            .filter(|term| term != &&Term::Terminal("ε".to_string()))
                        {
                            // First(Y1) ∖ {ε} to First(X)
                            changed |= self.insert_first_no_epsilon(&production.lhs, term);
                            println!(
                                "Rule3/4: Push First({}) \\ ε to First({})",
                                term, production.lhs
                            );
                            // terminate (check next expression) if X does NOT produce ε
                            if !self.produce_epsilon(term) {
                                println!("{} does NOT produce ε", term);
                                break;
                            }
                        }
                        // Rule 5: If X is a non-terminal and X -> Y1 Y2 ... Yk,
                        // and First(Yi) produce ε for all i, then add ε to First(X)
                        if expr.terms_iter().all(|term| self.produce_epsilon(term)) {
                            println!("Rule5: Push ε to First({})", production.lhs);
                            changed = self.insert_epsilon(&production.lhs);
                        }
                    }
                });

            if !changed {
                println!("Unchanged, break!");
                break;
            }
        } // End of loop
    }

    fn produce_epsilon(&self, term: &Term) -> bool {
        let production = self.lookup.get(&term);
        if production.is_none() {
            return false;
        }

        let production = production.unwrap();

        match &production.lhs {
            Term::Terminal(t) => {
                if t == "ε" {
                    return true;
                }
            }
            Term::Nonterminal(nt) => {
                if nt == "ε" {
                    return true;
                }
            }
        }

        production
            .rhs_iter()
            .map(|expr| {
                expr.terms_iter().all(|term| match term {
                    Term::Terminal(t) => t == "ε",
                    Term::Nonterminal(nt) => nt == "ε",
                })
            })
            .any(|v| v)
    }

    // Insert term to First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    fn insert_term(&self, x: &'grammar Term, term: &'grammar Term) -> bool {
        let mut first = self.first.borrow_mut();
        // First(x)
        dbg!(x);
        let first_x = first.get_mut(x).unwrap();

        // Insert term to First(x)
        let before = first_x.len();
        first_x.insert(term);
        let after = first_x.len();

        // check if set changes
        before != after
    }

    // Insert epsilon to First(x)
    fn insert_epsilon(&self, x: &'grammar Term) -> bool {
        self.insert_term(x, epsilon())
    }

    /// First(x)
    fn first(&self, x: &Term) -> HashSet<&'grammar Term> {
        self.first
            .borrow()
            .get(x)
            .map_or_else(HashSet::new, |set| set.clone())
    }

    fn insert_set(&self, x: &'grammar Term, set: HashSet<&'grammar Term>) -> bool {
        let mut first = self.first.borrow_mut();
        // First(x)
        let first_x = first.get_mut(x).unwrap();

        // Insert set into First(x)
        let before = first_x.len();
        first_x.extend(set);
        let after = first_x.len();

        // check if set changes
        before != after
    }

    /// Insert First(y) \ { ε } into First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    fn insert_first_no_epsilon(&self, x: &'grammar Term, y: &'grammar Term) -> bool {
        // First(y)
        let mut first_y = self.first(y);
        // First(y) \ { ε }
        first_y.remove(epsilon());
        // Insert First(y) \ { ε } into First(x)
        self.insert_set(x, first_y)
    }

    pub(crate) fn build(mut self) -> First<'grammar> {
        self.build_first();
        First {
            first: self.first.into_inner(),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::utils::first::builder::FirstBuilder;
    use bnf::Term;

    #[test]
    fn first() {
        let grammar = r#"
        <E> ::= <T> <E'>
        <E'> ::= '+' <T> <E'> | 'ε'
        <T> ::= <F> <T'>
        <T'> ::= '*' <F> <T'> | 'ε'
        <F> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();

        let first = FirstBuilder::new(&grammar).build();
        first.first.iter().for_each(|(lhs, rhs)| match lhs {
            Term::Terminal(_) => {
                assert_eq!(rhs.len(), 1)
            }
            Term::Nonterminal(_) => {
                assert_eq!(rhs.len(), 2)
            }
        })
    }
}
