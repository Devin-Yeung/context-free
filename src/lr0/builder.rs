use crate::lr0::core::{LR0Closure, LR0Item, LR0ItemSet};
use crate::utils::symbols;
use bnf::{Grammar, Production, Term};
use std::collections::{HashMap, VecDeque};

pub struct LR0Builder<'grammar> {
    grammar: &'grammar Grammar,
    closures: Vec<LR0ItemSet<'grammar>>,
    transitions: HashMap<(usize, &'grammar Term), usize>,
}

impl<'grammar> LR0Builder<'grammar> {
    pub fn new(grammar: &'grammar Grammar) -> LR0Builder<'grammar> {
        LR0Builder {
            grammar,
            closures: Vec::new(),
            transitions: HashMap::new(),
        }
    }

    pub fn build(mut self, augmentation: &'grammar Production) -> LR0Closure {
        let initial = LR0ItemSet::from_iter(vec![LR0Item::from_production(augmentation).unwrap()]);
        self.build_closure(&initial);
        self.build_transition();
        LR0Closure {
            closures: self.closures,
            transitions: self.transitions,
        }
    }

    fn build_closure(&mut self, initial: &LR0ItemSet<'grammar>) {
        let mut waiting = VecDeque::from([initial.closure(self.grammar)]);
        self.closures.push(initial.closure(self.grammar));

        while !waiting.is_empty() {
            let cur = waiting.pop_front().unwrap();
            symbols(self.grammar).iter().for_each(|term| {
                let goto = cur.goto(self.grammar, term);
                if !goto.items.is_empty() && !self.contains(&goto) {
                    self.closures.push(goto);
                    waiting.push_back(self.closures.last().unwrap().clone());
                }
            });
        }
    }

    fn build_transition(&mut self) {
        for closure in &self.closures {
            for term in symbols(self.grammar) {
                let goto = closure.goto(self.grammar, term);
                if goto.items.is_empty() {
                    continue;
                }
                let goto_index = self.index_of(&goto);
                let cur_index = self.index_of(closure);
                self.transitions.insert((cur_index, term), goto_index);
                println!("goto(I_{}, {}) = I_{}", cur_index, term, goto_index);
            }
        }
    }

    fn contains(&self, set: &LR0ItemSet<'grammar>) -> bool {
        self.closures.iter().any(|v| v == set)
    }

    fn index_of(&self, set: &LR0ItemSet<'grammar>) -> usize {
        self.closures.iter().position(|v| v == set).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use crate::lr0::core::LR0Closure;
    use bnf::Production;
    use std::str::FromStr;

    #[test]
    fn it_works() {
        let grammar = r#"
        <E'> ::= <E>
        <E> ::= <E> '+' <T> | <T>
        <T> ::= <T> '*' <F> | <F>
        <F> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();
        let augmentation = Production::from_str("<E'> ::= <E>").unwrap();

        assert_eq!(
            LR0Closure::new(&grammar, &augmentation).closures().len(),
            12
        );
    }

    #[test]
    fn object_clause() {
        let grammar = r#"
        <S> ::= <O> 'v' <C>
        <O> ::= 'n'
        <C> ::= <S>
        <C> ::= 'n'
        "#
        .parse()
        .unwrap();
        let augmentation = Production::from_str("<S'> ::= <S>").unwrap();

        assert_eq!(LR0Closure::new(&grammar, &augmentation).closures().len(), 8);
    }
}
