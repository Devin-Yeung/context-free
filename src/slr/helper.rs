use bnf::{Expression, Grammar, Production, Term};
use indexmap::IndexMap;
use itertools::Itertools;
use tabled::builder::Builder;
use tabled::Table;

pub struct IndexedGrammar<'grammar> {
    /// rhs -> lhs
    grammar: IndexMap<&'grammar Expression, &'grammar Term>,
    /// original
    original: &'grammar Grammar,
}

impl<'grammar> IndexedGrammar<'grammar> {
    pub fn new(grammar: &'grammar Grammar) -> IndexedGrammar<'grammar> {
        let original = grammar;
        let grammar = grammar
            .productions_iter()
            .flat_map(|production| production.rhs_iter().map(|expr| (expr, &production.lhs)))
            .collect::<IndexMap<&Expression, &Term>>();

        IndexedGrammar { grammar, original }
    }

    pub fn grammar_table(&self) -> Table {
        let mut builder = Builder::default();
        builder.push_record(["Rule"]);
        self.grammar.iter().for_each(|(rhs, lhs)| {
            builder.push_record([format!("{} -> {}", lhs, rhs)]);
        });
        builder.index().build()
    }

    pub fn get_index_of(&self, expr: &Expression) -> Option<usize> {
        self.grammar.get_index_of(expr)
    }

    pub(crate) fn get(&self, expr: &Expression) -> Option<Production> {
        let lhs = self.grammar.get(expr)?;
        Some(Production::from_parts(Term::clone(lhs), vec![expr.clone()]))
    }

    pub(crate) fn non_terminals<'a>(&'a self) -> impl Iterator<Item = &'grammar Term> {
        self.original
            .productions_iter()
            .map(|prod| &prod.lhs)
            .filter(|term| matches!(*term, Term::Nonterminal(_)))
            .unique()
    }

    pub(crate) fn terminals<'a>(&'a self) -> impl Iterator<Item = &'grammar Term> {
        self.original
            .productions_iter()
            .flat_map(|prod| prod.rhs_iter())
            .flat_map(|expr| expr.terms_iter())
            .filter(|term| matches!(*term, Term::Terminal(_)))
            .unique()
    }
}
