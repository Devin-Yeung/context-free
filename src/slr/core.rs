use crate::slr::helper::IndexedGrammar;
use crate::utils::dollar;
use bnf::Term;
use indexmap::IndexMap;
use itertools::Itertools;
use serde::ser::SerializeStruct;
use serde::{Serialize, Serializer};
use std::collections::HashMap;
use std::fmt::{Display, Formatter};
use std::iter::once;
use tabled::builder::Builder;
use tabled::Table;

#[derive(Clone, Serialize)]
pub enum SLRInstruction {
    Reduce(usize),
    Shift(usize),
    Goto(usize),
    Empty,
}

impl ToString for SLRInstruction {
    fn to_string(&self) -> String {
        match self {
            SLRInstruction::Reduce(i) => format!("r{}", i),
            SLRInstruction::Shift(i) => format!("s{}", i),
            SLRInstruction::Goto(i) => format!("g{}", i),
            SLRInstruction::Empty => String::new(),
        }
    }
}

impl From<&SLRInstruction> for String {
    fn from(val: &SLRInstruction) -> Self {
        val.to_string()
    }
}

pub struct SLRTable<'grammar> {
    pub(crate) grammar: IndexedGrammar<'grammar>,
    pub(crate) table: Vec<HashMap<&'grammar Term, SLRInstruction>>,
}

fn strip_quotes(s: &str) -> &str {
    if s.starts_with('"') && s.ends_with('"') && s.len() >= 2 {
        &s[1..s.len() - 1]
    } else {
        s
    }
}

impl<'grammar> Serialize for SLRTable<'grammar> {
    fn serialize<S>(&self, s: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut state = s.serialize_struct("SLRTable", 2)?;

        state.serialize_field("grammar", &self.grammar)?;

        let serialized_table: Vec<IndexMap<String, &SLRInstruction>> = self
            .table
            .iter()
            .map(|map| {
                map.iter()
                    .sorted_by_key(|(term, _)| *term)
                    .map(|(&term, instr)| (strip_quotes(&term.to_string()).to_string(), instr))
                    .collect::<IndexMap<_, _>>()
            })
            .collect();
        state.serialize_field("table", &serialized_table)?;
        state.end()
    }
}

impl<'grammar> SLRTable<'grammar> {
    pub fn grammar_table(&self) -> Table {
        self.grammar.grammar_table()
    }

    pub fn parsing_table(&self) -> Table {
        let mut builder = Builder::default();

        let header = self
            .grammar
            .terminals()
            .chain(once(dollar()))
            .chain(self.grammar.non_terminals())
            .collect::<Vec<_>>();

        builder.push_record(header.iter().map(|t| t.to_string()));

        self.table.iter().enumerate().for_each(|(_, table)| {
            let row = header
                .iter()
                .map(|t| table.get(t).unwrap_or(&SLRInstruction::Empty))
                .collect::<Vec<_>>();
            builder.push_record(row);
        });

        builder.index().build()
    }
}

impl<'grammar> Display for SLRTable<'grammar> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("Grammar: \n{}\n", self.grammar_table()))?;
        f.write_fmt(format_args!("Table: \n{}", self.parsing_table()))
    }
}

#[cfg(test)]
mod tests {
    use crate::slr::builder::SLRTableBuilder;
    use bnf::Production;
    use std::str::FromStr;

    #[test]
    fn slr_to_json() {
        let grammar = r#"
        <E'> ::= <E>
        <E> ::= <E> '+' <T> | <T>
        <T> ::= <T> '*' <F> | <F>
        <F> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();

        let augmentation = Production::from_str("<E'> ::= <E>").unwrap();

        let builder = SLRTableBuilder::new(&grammar, &augmentation);
        let slr = builder.build();
        let json = serde_json::to_string_pretty(&slr).unwrap();
        insta::assert_display_snapshot!(json);
    }
}
