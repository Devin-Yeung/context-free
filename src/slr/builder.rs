use crate::lr0::core::{LR0Closure, LR0Item};
use crate::slr::core::SLRInstruction;
use crate::slr::helper::IndexedGrammar;
use crate::utils::dollar;
use crate::utils::follow::Follow;
use bnf::{Grammar, Production, Term};
use std::cell::RefCell;
use std::collections::HashMap;
use std::fmt::{format, Display, Formatter};
use std::iter::{once, repeat};
use tabled::builder::Builder;
use tabled::Table;

pub struct SLRTable<'grammar> {
    grammar: IndexedGrammar<'grammar>,
    table: Vec<HashMap<&'grammar Term, SLRInstruction>>,
}

impl<'grammar> SLRTable<'grammar> {
    pub fn grammar_table(&self) -> Table {
        self.grammar.grammar_table()
    }

    pub fn parsing_table(&self) -> Table {
        let mut builder = Builder::default();

        let header = self
            .grammar
            .terminals()
            .chain(once(dollar()))
            .chain(self.grammar.non_terminals())
            .collect::<Vec<_>>();

        builder.set_header(header.iter().map(|t| t.to_string()));

        self.table.iter().enumerate().for_each(|(_, table)| {
            let row = header
                .iter()
                .map(|t| table.get(t).unwrap_or(&SLRInstruction::Empty))
                .collect::<Vec<_>>();
            builder.push_record(row);
        });

        builder.index().build()
    }
}

impl<'grammar> Display for SLRTable<'grammar> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("Grammar: \n{}\n", self.grammar_table()))?;
        f.write_fmt(format_args!("Table: \n{}", self.parsing_table()))
    }
}

pub struct SLRTableBuilder<'grammar> {
    grammar: IndexedGrammar<'grammar>,
    follow: Follow<'grammar>,
    closure: LR0Closure<'grammar>,
    table: RefCell<Vec<HashMap<&'grammar Term, SLRInstruction>>>,
}

impl<'grammar> SLRTableBuilder<'grammar> {
    pub fn new(
        grammar: &'grammar Grammar,
        augmentation: &'grammar Production,
    ) -> SLRTableBuilder<'grammar> {
        let follow = Follow::new(grammar, &augmentation.lhs);
        let closure = LR0Closure::new(grammar, augmentation);
        let grammar = IndexedGrammar::new(grammar);
        let table = RefCell::new(
            repeat(HashMap::<&Term, SLRInstruction>::new())
                .take(closure.len())
                .collect(),
        );
        SLRTableBuilder {
            grammar,
            follow,
            closure,
            table,
        }
    }

    fn shift(&self, from: usize, via: &'grammar Term) {
        let to = self.closure.transition(from, via).unwrap(); // TODO: really not sure whether this unwrap is safe?
        println!("Shift: goto(I_{}, {}) = I_{}", from, via, to);
        let mut table = self.table.borrow_mut();
        table
            .get_mut(from)
            .unwrap()
            .insert(via, SLRInstruction::Shift(to));
    }

    fn reduce(&self, index: usize, lr0: &LR0Item) {
        let grammar_index = self.grammar.get_index_of(&lr0.rhs).unwrap();
        let prod = self.grammar.get(lr0.rhs).unwrap();
        let mut table = self.table.borrow_mut();
        for term in self.follow.follow_of(&prod.lhs).collect::<Vec<_>>() {
            table[index].insert(term, SLRInstruction::Reduce(grammar_index));
            println!(
                "Reduce: set (I_{}, {}) = r{}",
                index,
                term.to_string(),
                grammar_index
            );
        }
    }

    fn goto(&self, from: usize, via: &'grammar Term) {
        debug_assert!(matches!(via, Term::Nonterminal(_)));
        let to = self.closure.transition(from, via).unwrap(); // TODO: really not sure whether this unwrap is safe?
        println!("Goto: goto(I_{}, {}) = I_{}", from, via, to);
        let mut table = self.table.borrow_mut();
        table
            .get_mut(from)
            .unwrap()
            .insert(via, SLRInstruction::Goto(to));
    }

    pub fn build(self) -> SLRTable<'grammar> {
        self.closure.enumerate_lr0().for_each(|(i, lr0)| {
            // take a look at the char after the •
            match lr0.expect() {
                // lr0 expect non character, which is the form of A -> 𝛼 •
                None => {
                    /* Reduce */
                    self.reduce(i, lr0)
                }
                Some(t) => {
                    match t {
                        Term::Terminal(_) => self.shift(i, t),   /* Shift */
                        Term::Nonterminal(_) => self.goto(i, t), /* Goto */
                    }
                }
            }
        });
        SLRTable {
            grammar: self.grammar,
            table: self.table.into_inner(),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::slr::builder::SLRTableBuilder;
    use bnf::Production;
    use std::str::FromStr;

    #[test]
    fn it_works() {
        let grammar = r#"
        <E'> ::= <E>
        <E> ::= <E> '+' <T> | <T>
        <T> ::= <T> '*' <F> | <F>
        <F> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();

        let augmentation = Production::from_str("<E'> ::= <E>").unwrap();

        let builder = SLRTableBuilder::new(&grammar, &augmentation);
        let slr = builder.build();
        insta::assert_display_snapshot!(slr);
    }
}
