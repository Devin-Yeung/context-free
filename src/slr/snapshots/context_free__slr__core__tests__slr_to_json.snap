---
source: src/slr/core.rs
expression: json
---
{
  "grammar": [
    {
      "lhs": {
        "Nonterminal": "E'"
      },
      "rhs": [
        {
          "terms": [
            {
              "Nonterminal": "E"
            }
          ]
        }
      ]
    },
    {
      "lhs": {
        "Nonterminal": "E"
      },
      "rhs": [
        {
          "terms": [
            {
              "Nonterminal": "E"
            },
            {
              "Terminal": "+"
            },
            {
              "Nonterminal": "T"
            }
          ]
        },
        {
          "terms": [
            {
              "Nonterminal": "T"
            }
          ]
        }
      ]
    },
    {
      "lhs": {
        "Nonterminal": "T"
      },
      "rhs": [
        {
          "terms": [
            {
              "Nonterminal": "T"
            },
            {
              "Terminal": "*"
            },
            {
              "Nonterminal": "F"
            }
          ]
        },
        {
          "terms": [
            {
              "Nonterminal": "F"
            }
          ]
        }
      ]
    },
    {
      "lhs": {
        "Nonterminal": "F"
      },
      "rhs": [
        {
          "terms": [
            {
              "Terminal": "("
            },
            {
              "Nonterminal": "E"
            },
            {
              "Terminal": ")"
            }
          ]
        },
        {
          "terms": [
            {
              "Terminal": "id"
            }
          ]
        }
      ]
    }
  ],
  "table": [
    {
      "(": {
        "Shift": 1
      },
      "id": {
        "Shift": 2
      },
      "<E>": {
        "Goto": 3
      },
      "<F>": {
        "Goto": 4
      },
      "<T>": {
        "Goto": 5
      }
    },
    {
      "(": {
        "Shift": 1
      },
      "id": {
        "Shift": 2
      },
      "<E>": {
        "Goto": 6
      },
      "<F>": {
        "Goto": 4
      },
      "<T>": {
        "Goto": 5
      }
    },
    {
      "$": {
        "Reduce": 6
      },
      ")": {
        "Reduce": 6
      },
      "*": {
        "Reduce": 6
      },
      "+": {
        "Reduce": 6
      }
    },
    {
      "$": {
        "Reduce": 0
      },
      "+": {
        "Shift": 7
      }
    },
    {
      "$": {
        "Reduce": 4
      },
      ")": {
        "Reduce": 4
      },
      "*": {
        "Reduce": 4
      },
      "+": {
        "Reduce": 4
      }
    },
    {
      "$": {
        "Reduce": 2
      },
      ")": {
        "Reduce": 2
      },
      "*": {
        "Shift": 8
      },
      "+": {
        "Reduce": 2
      }
    },
    {
      ")": {
        "Shift": 9
      },
      "+": {
        "Shift": 7
      }
    },
    {
      "(": {
        "Shift": 1
      },
      "id": {
        "Shift": 2
      },
      "<F>": {
        "Goto": 4
      },
      "<T>": {
        "Goto": 10
      }
    },
    {
      "(": {
        "Shift": 1
      },
      "id": {
        "Shift": 2
      },
      "<F>": {
        "Goto": 11
      }
    },
    {
      "$": {
        "Reduce": 5
      },
      ")": {
        "Reduce": 5
      },
      "*": {
        "Reduce": 5
      },
      "+": {
        "Reduce": 5
      }
    },
    {
      "$": {
        "Reduce": 1
      },
      ")": {
        "Reduce": 1
      },
      "*": {
        "Shift": 8
      },
      "+": {
        "Reduce": 1
      }
    },
    {
      "$": {
        "Reduce": 3
      },
      ")": {
        "Reduce": 3
      },
      "*": {
        "Reduce": 3
      },
      "+": {
        "Reduce": 3
      }
    }
  ]
}
